twitter_api/main.py

```diff
@@ -1,28 +1,217 @@
 import sqlite3
-
-from flask import Flask
+import json
+from flask import Flask, Response, request, abort, url_for
 from flask import g
 
-app = Flask(__name__)
+# extra modules we added
+from hashlib import md5
+import random
+from .utils import JSON_MIME_TYPE, auth_only, json_only
+import string
+from datetime import datetime
 
+app = Flask(__name__)
 
 def connect_db(db_name):
     return sqlite3.connect(db_name)
 
-
 @app.before_request
 def before_request():
     g.db = connect_db(app.config['DATABASE'])
-
-
+    
 # implement your views here
+def _get_tweet_info(tweet_id):
+    # returns (tweet_id, user_id, created, content)
+    sql_command = 'SELECT * FROM tweet WHERE id = ?'
+    tweet_cursor = g.db.execute(sql_command, [tweet_id])
+    tweet = tweet_cursor.fetchone()
+    if not tweet:
+        abort(404)
+    return tweet
+    
+def _get_user_id_from_token(request):
+    # assumes access token has already been verified but can double check
+    if 'access_token' not in request.get_json():
+        abort(401)
+    access_token = request.get_json()['access_token']
+    sql_command = 'SELECT user_id FROM auth WHERE access_token = ?'
+    user_id_cursor = g.db.execute(sql_command, [access_token])
+    found_user = user_id_cursor.fetchone()
+    if not found_user:
+        abort(404)
+    return found_user[0]
+    
+def _convert_date_iso_8601(created):
+    dt_object = datetime.strptime(created, "%Y-%m-%d %H:%M:%S")
+    return dt_object.strftime("%Y-%m-%dT%H:%M:%S")
+    
+@app.route('/login', methods = ['POST'])
+@json_only
+def login():
+    data = request.get_json()
+    
+    if 'username' not in data:
+        abort(400) # Bad request, no username
+    elif 'password' not in data:
+        abort(400) # Bad request, no password
 
+    post_username = data['username']
+    post_password = data['password']
+    
+    sql_command = 'SELECT password, id FROM user WHERE username = ?'
+    
+    found_user_cursor = g.db.execute(sql_command, [post_username])
+    found_user = found_user_cursor.fetchone()
+    if not found_user:
+        abort(404) # Not found, username did find a match in db
+    password, id = found_user
+        
+    post_password = post_password.encode(request.charset) # note from Santiago
+    post_password = md5(post_password).hexdigest()
+    
+    if password != post_password:
+        abort(401) # Unauthorized
+        
+    # generate access token
+    char_list = string.ascii_lowercase + string.ascii_uppercase
+    access_token = ''.join(random.choice(char_list) for _ in range(10))
```

Comment:
So, purely being pedantic here, whilst there are 56! (8.1x10e67) possible 10 character access tokens based on your char_list, there is the non zero, though almost infinitesimally small, possibility that two of the users of your API could end up with the same access token.
To avoid this I would either append something unique to the user to the 10 character random list you generate:
```python
access_token = str(username).join(random.choice(char_list) for _ in range(10))
```
This would be a guaranteed way to ensure no two users, who by some miracle had the same random 10 character string generated, could have the same access token.
alternatively if you wanted to obscure the username you could do something like:
```
access_token = md5(username.join(random.choice(char_list) for _ in range(10))
```